utils/neural_magic/sparse_train_manager.py

```diff
@@ -0,0 +1,208 @@
+        )
+
+        # If recipe contains lr modifiers, turn off native lr scheduler
+        if self.train_manager.learning_rate_modifiers:
+            scheduler = None
+
+        # If recipe contains epoch range modifiers, overwrite epoch range
+        if self.train_manager.epoch_modifiers and self.train_manager.max_epochs:
+            epochs = self.train_manager.max_epochs
+
+        # construct a ToggleableModelEMA from ModelEMA, allowing for disabling for QAT
+        ema = load_ema(ema.ema.state_dict(), self.model, **ema_kwargs)
+
+        return scaler, scheduler, ema, epochs
+
+    def qat_active(self, epoch: int) -> bool:
+        """
+        Returns true if QAT is turned on for the given epoch
+
+        :param epoch: epoch to check QAT status for
+        """
+        if self.train_manager.quantization_modifiers:
+            qat_start = min(
+                [mod.start_epoch for mod in self.train_manager.quantization_modifiers]
+            )
+            return qat_start < epoch + 1
+        else:
+            return False
+
+    def is_qat_recipe(self) -> bool:
+        """
+        Returns true if the training recipe contains a QAT modifier
+        """
+        return bool(self.train_manager.quantization_modifiers)
+
+    def turn_off_scaler(self, scaler: torch.cuda.amp.GradScaler):
+        """
+        Turns off grad scaler
+
+        :param scaler: scaler to run off
+        """
+        scaler._enabled = False
+
+    def update_state_dict_for_saving(
+        self, ckpt: Dict[str, Any], final_epoch: bool, ema_enabled: bool
+    ) -> Dict[str, Any]:
+        """
+        Update checkpoint dictionary to be compatible with sparse model saving
+
+        :param ckpt: original checkpoint dictionary
+        :param final_epoch: True if called after last training epoch
+        :param ema_enabled: True if ema is turned on
+        """
+        if final_epoch:
+            # save model with a checkpoint recipe representing all recipes applied to
+            # model, allowing for multiple stages of sparse training
+            checkpoint_recipe = (
+                ScheduledModifierManager.compose_staged(
+                    self.checkpoint_manager, self.train_manager
+                )
+                if self.checkpoint_manager
+                else self.train_manager
+            )
+        else:
+            checkpoint_recipe = None
+
+        # Pickling is not supported for quantized models for a subset of the supported
+        # torch versions, thus models are saved via their state dict for all sparse
+        # models
+        sparseml_dict_update = {
+            "model": ckpt["model"].state_dict(),
+            "yaml": ckpt["model"].yaml,
+            "ema": ckpt["ema"].state_dict() if ema_enabled else None,
+            "updates": ckpt["updates"] if ema_enabled else None,
+            "checkpoint_recipe": str(checkpoint_recipe),
+        }
+        ckpt.update(sparseml_dict_update)
```

Comment:
is epoch in this checkpoint? do you need to update epoch if final_epoch == True?

Reply:
I think this could go either way. I added it for now, but will see how it plays out with the other changes. Thanks for reminder!